lms/djangoapps/commerce/utils.py

```diff
@@ -109,3 +117,263 @@ def upgrade_url(self, user, course_key):
+        log.info('Refund successfully opened for user [%s], course [%s]: %r', enrollee.id, course_key_str, refund_ids)
+
+        _process_refund(
+            refund_ids=refund_ids,
+            api_client=api_client,
+            course_product=course_enrollment,
+        )
+    else:
+        log.info('No refund opened for user [%s], course [%s]', enrollee.id, course_key_str)
+
+    return refund_ids
+
+
+def _process_refund(refund_ids, api_client, course_product):
+    """
+    Helper method to process a refund for a given course_product. This method assumes that the User has already
+    been unenrolled.
+
+    Returns:
+        bool: True if the refund process was successful, False if there are any Errors that are not handled
+    """
+    config = CommerceConfiguration.current()
+
+    if config.enable_automatic_refund_approval:
+        refunds_requiring_approval = []
+
+        for refund_id in refund_ids:
+            try:
+                # NOTE: The following assumes that the user has already been unenrolled.
+                # We are then able to approve payment. Additionally, this ensures we don't tie up an
+                # additional web worker when the E-Commerce Service tries to unenroll the learner.
+                api_client.refunds(refund_id).process.put({'action': 'approve_payment_only'})
+                log.info('Refund [%d] successfully approved.', refund_id)
+            except:  # pylint: disable=bare-except
+                # Push the refund to Support to process
+                log.exception('Failed to automatically approve refund [%d]!', refund_id)
+                refunds_requiring_approval.append(refund_id)
+    else:
+        refunds_requiring_approval = refund_ids
+
+    if refunds_requiring_approval:
+        # XCOM-371: this is a temporary measure to suppress refund-related email
+        # notifications to students and support for free enrollments.  This
+        # condition should be removed when the CourseEnrollment.refundable() logic
+        # is updated to be more correct, or when we implement better handling (and
+        # notifications) in Otto for handling reversal of $0 transactions.
+        if course_product.mode != 'verified':
+            # 'verified' is the only enrollment mode that should presently
+            # result in opening a refund request.
+            msg = 'Skipping refund support notification for non-verified mode for user [%s], course [%s], mode: [%s]'
+            course_identifier = course_product.course_id
+            if course_product.uuid:
+                course_identifier = str(course_product.uuid)
+                msg = ('Skipping refund support notification for non-verified mode for user [%s], '
+                       'course entitlement [%s], mode: [%s]')
+            log.info(
+                msg,
+                course_product.user.id,
+                course_identifier,
+                course_product.mode,
+            )
+            return False
```

Comment:
Maybe leave a comment clarifying that this will result in a UI error for entitlements, but not for seats.

Reply:
Actually, I'm not convinced that returning `False` here is correct. The docstring says that `False` indicates that something went wrong, but I'm not sure that's actually the case here, unless the course product is an entitlement.

Reply:
I think entitlements should not hit this code block at all. I cannot think of a reason why we wouldn't want to send notifications to support for entitlement refund requests. Perhaps we could add a check for whether or not the course product is an entitlement to https://github.com/edx/edx-platform/pull/16934/files#diff-4394a757d8d82dd3dc4351b59a073065R261 ?

Reply:
I do not think that is necessary as that is a nuance of entitlements.

Reply:
Lets talk about this tomorrow. I think it's incorrect to always return `False` here as that implies something went wrong, when for most non-entitlement refund requests that is not the case.

Reply:
The return value is not currently used by non-entitlement refunds. I will remove the False return here as it will not break the process at all and if one gets through it will just sit in awaiting approval.